overlay: warn if overlay backing fs doesn't support d_type

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**

Print a warning if the backing fs for overlay/overlay2 doesn't support `d_type`.
Close #27358 

Considering the discussion at https://github.com/docker/docker/issues/22937, the warning is not treated as an error.
However, I still think it is useful to treat it as an error in some future version (1.14? 1.15?).

Note that recent kernel prints a similar message: https://github.com/torvalds/linux/commit/e7c0b5991dd1be7b6f6dc2b54a15a0f47b64b007
However I think there are advantages to print the message as well in Docker:
- Can support older kernels
- Can provide much more user-friendly information (e.g. "You need to reformat with f_type=1" for XFS)
- Can be switched to an error in future Docker versions

It also shows information in `docker info`.

**\- How I did it**
Please refer to the source of `pkg/fsutils.SupportsDType`.

**\- How to verify it**
For example, XFS with ftype=0 prints this warning.

``` console
$ qemu-img create /xfs-ftype0.img 10G
$ mkfs.xfs -m crc=0 -n ftype=0 /xfs-ftype0.img
$ mount -o loop /xfs-ftype0.img /mnt
$ dockerd -s overlay2 -g /mnt
WARN[0001] overlay2 won't work properly because the backing filesystem xfs is not configured to support d_type. (The backing filesystem seems formatted with f_type=0. Please r
eformat the filesytem with f_type=1.) Note that this warning is planned to be treated as a fatal error in v1.16 

```

The warning should not be shown if ftype=1.

`docker info` shows `f_type Supported` information:

``` console
$ docker info
...
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: false
...
```

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**\- A picture of a cute animal (not mandatory but encouraged)**

TODO
- [ ] PR for doc (after this PR gets merged)

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

AkihiroSuda committed:
overlay: warn if overlay backing fs doesn't support d_type

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented:
CC @rhvgoyal

thaJeztah commented:
/cc @dmcgowan @justincormack

rhvgoyal commented:
If this is just a warning, then kernel is already emitting a warning. How does it help doing another one.

dmcgowan commented:
> How does it help doing another one.

This was prompted by the issue here #27358. The warning should have been shown in the kernel but not is not somewhere users will look before filing an issue. However I am not convinced that adding a warning to the Docker logs will make this situation much better.

@AkihiroSuda what if instead we added a key mentioning d_type status to the the [status output](https://github.com/docker/docker/blob/master/daemon/graphdriver/overlay/overlay.go#L176) so when issues are filed it is easier to see the problem in the `docker info` output without making users dump their logs. My feeling here is we have to add something to make diagnosis easier, mentioning in the documentation and having warning in the logs can only go so far. I also don't see it as reasonable to someday flip this to an error, the kernel update to make this a warning was done because it had an impact on users, how would that impact change in the future?

thaJeztah commented:
@dmcgowan I guess we should decide how serious the situation is if someone runs docker without this. If it's a serious issue, then the `overlay` driver should probably be skipped (as "unsupported"), otherwise, we could output it as warning in `docker info` (where we already output various warnings if things are not supported). wdyt?

AkihiroSuda commented:
Updated PR

* Now warning emits user-friendly guidance for xfs ("The backing filesystem seems formatted with f_type=0. Please reformat with f_type=1.")
* Now `docker info` contains `f_type Supported` information

@rhvgoyal 
* This PR can support older kernel that doesn't print warning
* This PR can provide user-friendly guidance ("The backing filesystem seems formatted with f_type=0. Please reformat with f_type=1.")

@dmcgowan 
Makes sense, I added the information to `docker info`

@thaJeztah 
IMO overlay is almost unuseful without d_type, and it should be an error after some "grace period" (1.14 or 1.15)

thaJeztah commented:
> IMO overlay is almost unuseful without d_type, and it should be an error after some "grace period" (1.14 or 1.15)

If we are ok with doing that, we should add that in the warning ("running without t_type=0 will no longer be supported in 1.15" for example (needs a better sentence))

AkihiroSuda commented:
If this should be considered as a "feature deprecation", the warning should remain in v1.13, v1.14. and v.1.15. So the warning will be "running without t_type=0 will no longer be supported in 1.16" in such a case. 
https://docs.docker.com/engine/#feature-deprecation-policy
> Before an existing feature is removed it will be labeled as “deprecated” within the documentation and will remain in Docker for, usually, at least 3 releases. After that time it may be removed.

IMO this is not a "feature deprecation", because "supporting overlay without  d_type"  is a bug rather than a feature.
However, I'm ok to set the grace period (when the warning is changed to an error) to v1.16 so that it won't surprise users.

thaJeztah commented:
We discussed this, and suggest to:

- make this show up in docker info
- show a warning if it's not supported (as suggested above)
- make this a "fatal" error in 1.16

This allows users that currently run without this being supported to migrate/address the issue

tonistiigi commented on pkg/fsutils/fsutils.h:
```diff
@@ -0,0 +1,29 @@
+  }
+  struct dirent *ent;
+  while ((ent = readdir(d)) != NULL) {
+    if (ent->d_type == DT_UNKNOWN) {
```
It should be possible to do this without c. https://golang.org/pkg/syscall/#Dirent is exposed in go.

AkihiroSuda replied:
done

AkihiroSuda commented:
@thaJeztah thank you for making the discussion. 

Updated PR and now the warning message is like this:
```
WARN[0001] overlay won't work properly because the backing filesystem xfs is not configured to support d_type. (The backing filesystem seems formatted with f_type=0. Please r
eformat the filesytem with f_type=1.) Note that this warning is planned to be treated as a fatal error in v1.16 .
```

thaJeztah commented:
Trying to see if we can make this more consistent with the devmapper dev-sync message;

```
devmapper: Udev sync is not supported. This will lead to data loss and unexpected
behavior. Install a dynamic binary to use devicemapper or select a different storage
driver. For more information, see https://docs.docker.com/engine/reference/commandline/daemon/#daemon-storage-driver-option

```

Perhaps (wrapped for readability):

```
WARN[0001] overlay2: the backing xfs filesystem is configured without d_type support,
which leads to incorrect behavior. Reformat the filesytem with f_type=1 to fix this
issue. Running without d_type support will no longer be supported in Docker 1.16.
```

@mstanleyjones perhaps you have suggestions?

AkihiroSuda commented:
Updated PR according to the suggestion from @mstanleyjones

mdlinville reviewed (approved):
Docs LGTM

thaJeztah commented:
ping @justincormack @tonistiigi PTAL

dmcgowan commented on pkg/fsutils/fsutils_linux.go:
```diff
@@ -0,0 +1,90 @@
+			// stop walking
+			return true
+		}
+		// continue walking
```
What is the reasoning behind checking multiple files? Is there a case where there can be a mix of unknown type and known types?

AkihiroSuda replied:
I agree that checking a single file is enough in practice.

But I wonder theoretically there can be a mix of unknown type and known types in some filesystem, so I wrote the code to check multiple files here. ( it should be very lightweight operation)

Please let me know if I should fix this.

AkihiroSuda commented on pkg/fsutils/fsutils_linux.go:
```diff
@@ -0,0 +1,90 @@
+	return supportsDType, nil
+}
+
+func walkdir(path string, walkFn func(*syscall.Dirent) bool) error {
```
self-nit: `walk` maybe misleading because it is not recursive.
I'll try to change the terminology

AkihiroSuda replied:
done

AkihiroSuda commented:
rebased to master
@justincormack @tonistiigi PTAL

vieux commented:
ping @justincormack @tonistiigi

AkihiroSuda commented on pkg/fsutils/fsutils_linux_test.go:
```diff
@@ -0,0 +1,90 @@
+
+// prepare creates an empty sparse image and a directory for mounting,
+// under tmpDir (can be empty)
+func prepare(t *testing.T, tmpDir string, imageSize int64) (string, string) {
```
done

tonistiigi commented:
LGTM

dmcgowan commented:
LGTM
